### PR TITLE
Adding support for tracking Motherboard and BIOS firmware information

### PR DESCRIPTION
**Pull request by jyundt**
I'm open to suggestions on the best way to track this information, but here is an initial stab.

Screenshot of the updated Server Base hardware details view:

image

I added four tests to validate the motherboard/BIOS info is correct. Let me know if there are any other edge cases that would be appropriate.

(For some background our hardware teams are going to start using this information and we'd like to get it into collins.)

Fixes tumblr/collins#555

cc @michaeljs1990 @byxorna

### CODE REVIEW

**Comment by jyundt** [comment]
cc @nshalman

**Comment by nshalman** [comment]
IIRC someone from hardware suggested tracking the firmware build date as well.

**Comment by jyundt** [comment]
> IIRC someone from hardware suggested tracking the firmware build date as well.

Yeah I _just_ realized I didn't include that. I'll update this shortly.

**Comment by byxorna** [comment]
I am not sure where the line lies, but at some point collecting "facts" about your hardware during induction in genesis and setting normal attributes may be less resistance than needing to extract info from LSHW and bake straight into collins' hardware details. For bios/board model/version, its definitely useful to have in the hardware details. For other, more esoteric fields, I'd err more on the you-set-whatever-attributes-you-want-during-induction side :P

File: app/collins/models/lshw/ServerBase.scala [hunk 1]
```diff
@@ -22,9 +26,9 @@ object ServerBase {
 }
 
 case class ServerBase(
-    description: String = "", product: String = "", vendor: String = "", serial: Option[String] = None) extends LshwAsset {
+    motherboard: String = "", firmware: String = "", description: String = "", product: String = "", vendor: String = "", serial: Option[String] = None) extends LshwAsset {
```
Comment (byxorna) [inline app/collins/models/lshw/ServerBase.scala]: why not make these options instead of strings with empty defaults?

File: app/collins/util/parsers/LshwParser.scala [hunk 2]
```diff
@@ -197,13 +197,29 @@ class LshwParser(txt: String) extends CommonParser[LshwRepresentation](txt) {
       val asset = getAsset(elem)
       // serial may be missing, so be flexible here and allow it to be absent
       val serial = (elem \ "serial" headOption).map(_.text)
-      ServerBase(asset.description, asset.product, asset.vendor, serial)
+      val motherboard = (elem \ "node" \ "product").headOption match {
```
Comment (byxorna) [inline app/collins/util/parsers/LshwParser.scala]: using Option[String] would simplify this boilerplate logic (/see line 199)

File: app/views/asset/show_hwdetails.scala.html [hunk 3]
```diff
@@ -26,10 +26,15 @@
...
         </tr>
-
         <tr>
             <td>Serial</td><td>@aa.lshw.base.serial</td>
         </tr>
+        <tr>
+            <td>Motherboard</td><td>@aa.lshw.base.motherboard</td>
+        </tr>
+        <tr>
+            <td>Firmware</td><td>@aa.lshw.base.firmware</td>
```
Comment (byxorna) [inline app/views/asset/show_hwdetails.scala.html]: its not clear if this is firmware for mobo, ipmi module, bios, or some other bit. Can you clarify, maybe in a `title=""`?

Comment (jyundt) [reply app/views/asset/show_hwdetails.scala.html]: I agree, it's not very descriptive. 

I might play around with this a bit more:

image

File: app/collins/util/parsers/LshwParser.scala (hunk 2 above)
Comment (jyundt) [reply app/collins/util/parsers/LshwParser.scala]: Sounds good to me, changed in e64829b54711877d6342e15b8e9019a4b0440eb8.

File: app/collins/models/lshw/ServerBase.scala (hunk 1 above)
Comment (jyundt) [reply app/collins/models/lshw/ServerBase.scala]: Sounds good to me, changed in e64829b54711877d6342e15b8e9019a4b0440eb8.

**Comment by byxorna** [comment]
@jyundt i like it. anything else that would be generally useful that we should get into this evolution before landing this? cc @roymarantz @defect @gtorre @qx-xp

File: conf/evolutions/collins/15.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# --- Add GPU information to asset_meta
```
Comment (defect) [inline conf/evolutions/collins/15.sql]: 👀  Mind updating this to match what it does? :)

**Comment by defect** [comment]
Only took a quick glance but i like it! I'll fire it up tomorrow and try it out a little more closely.

File: conf/evolutions/collins/15.sql (hunk 4 above)
Comment (jyundt) [reply conf/evolutions/collins/15.sql]: :man_facepalming: 

Good catch.

**Comment by byxorna** [comment]
@defect lets land this?

**Comment by defect** [comment]
Yeah, definitely. I wanted to run the LSHW output from a few types of hardware through it to see what happened, and of course forgot about it. I'll do it first thing tomorrow (don't have my work laptop with me) and merge it.
